src/ros2idl.ne

```diff
@@ -0,0 +1,395 @@
+%}
+
+assignment -> (
+    floatAssignment
+  | intAssignment
+  | stringAssignment
+  | booleanAssignment
+) {% d => d[0][0] %}
+
+floatAssignment ->   _ %EQ _ (SIGNED_FLOAT | FLOAT) {% ([_, __, ___,  num]) => ({valueText: num[0], value: parseFloat(num[0])}) %}
+intAssignment -> _ %EQ _ (SIGNED_INT | INT) {% ([_, __,___,  num]) => ({valueText: num[0], value: parseInt(num[0])}) %}
+stringAssignment -> _ %EQ _ STR {% ([_, __, ___, str]) => ({valueText: str, value: str}) %}
+booleanAssignment -> _ %EQ _ BOOLEAN {% ([_, __, ___, bool]) => ({valueText: bool, value: bool === "TRUE"}) %}
+
+allTypes -> (
+    primitiveTypes
+  | customType
+) {% d => d[0][0] %}
+
+primitiveTypes -> (
+    stringType
+  | numericType
+  | booleanType
+) {% d => ({...d[0][0], isComplex: false}) %}
+
+customType -> %NAME {% d => ({type: d[0].value, isComplex: true }) %}
+
+stringType ->  ("string"|"wstring") (_ "<" _ (INT | %NAME) _ ">"):? {% d => {
+	let strLength = undefined;
+	if(d[1] !== null) {
+		strLength = getIntOrConstantValue(d[1][3] ? d[1][3][0] : undefined);
+	}
+	return {type: "string", upperBound: strLength};
+} %}
+
+booleanType -> "boolean" {% d => ({type: "bool"}) %}
+
+# order matters here
+numericType -> (
+    "byte"
+  | "octet"
+  | "wchar"
+  | "char"
+  | "long" __ "double"
+  | "double"
+  | "float"
+  | "int8"
+  | "uint8"
+  | "int16"
+  | "uint16"
+  | "int32"
+  | "uint32"
+  | "int64"
+  | "uint64"
+  | "unsigned" __ "short"
+  | "short"
+  | "unsigned" __ "long" __ "long"
+  | "long" __ "long"
+  | "unsigned" __ "long"
+  | "long" 
+) {% function(d, reject) { 
+
+	const maybeType = d[0].map((t) => t?.value).filter(t => !!t).join(" ");
+	switch(maybeType) {
+		case "unsigned short":
+			return {type: "uint16"};
+			
+		case "unsigned long":
+			return {type: "uint32"};
+			
+		case "unsigned long long":
+			return {type: "uint64"};
+			
+		case "short":
+			return {type: "int16"};
+			
+		case "long":
+			return {type: "int32"};
+			
+		case "long long":
+			return {type: "int64"};
+			
+		case "double":
+		case "float":
+			return {type: "float32"};
+			
+		case "long double":
+			return {type: "float64"};
+			
+		case "octet":
+			return {type: "byte"};
+		
+		case "wchar": 
+			return {type: "char"};
+
+		default:
+			return { type: maybeType } ;
+	}
+}
+%}
+
+# ALL CAPS return strings rather than objects (terminals)
+
+BOOLEAN -> ("TRUE" | "FALSE") {% join %}
+
+# need to support mutliple adjacent strings as a single string
+STR -> (%STRING _):+  {% d => {
+	return join(d[0].flat(1).filter(d => d !== null));
+}%}
+
+# Not actually used due to needing to parse either an int or float from a string
+NUMBER -> (SIGNED_FLOAT | SIGNED_INT | FLOAT | INT)  {% join %}
+
+SIGNED_FLOAT -> ("+"|"-") FLOAT {% join %}
+
+FLOAT -> (%DECIMAL|%DECIMALEXP) {% join %}
+ | (%DECIMAL "d") {% d => d[0][0].value %}
+ | (INT "d") {% d => d[0][0] %}
+
+
+SIGNED_INT -> ("+"|"-") INT  {% join %}
+
+INT -> %INTEGER {% join %}
```

Comment:
what's the point of this rule?

Reply:
converts the token object to string so that it's a little easier to work with. will comment